test/xds/xds_client_ignore_resource_deletion_test.go

```diff
@@ -0,0 +1,474 @@
+			return
+		case <-timer.C:
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// This subtest tests the scenario where the bootstrap config has "ignore_resource_deletion"
+// not set in "server_features" field. This subtest verifies that the resource was
+// deleted by the xDSClient when a resource is missing the xDS response and subsequent
+// RPCs fail.
+func testResourceDeletionNotIgnored(t *testing.T, resources e2e.UpdateOptions, u func(r *e2e.UpdateOptions)) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout*1000)
+	t.Cleanup(cancel)
+	ms := startManagementServer(t)
+	bootstrapContent := generateBootstrapContents(t, ms.Address, false)
+
+	// Update the management server with initial resources setup.
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	cResolver := generateCustomResolver(t, bootstrapContent)
+	cc, err := grpc.Dial(fmt.Sprintf("xds:///%s", serviceName), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(cResolver))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	verifyRPCtoAllEndpoints(t, cc)
+
+	// Mutate resource and update on the server.
+	u(&resources)
+	if err := ms.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Spin up go routines to verify RPCs fail after the update.
+	client := testServicepb.NewTestServiceClient(cc)
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	go func() {
+		for ctx.Err() == nil {
+			if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+				wg.Done()
+				break
+			}
+			time.Sleep(50 * time.Millisecond)
+		}
+	}()
+	go func() {
+		for ctx.Err() == nil {
+			if _, err := client.UnaryCall(ctx, &testpb.SimpleRequest{}); err != nil {
+				wg.Done()
+				break
+			}
+			time.Sleep(50 * time.Millisecond)
+		}
+	}()
+
+	wg.Wait()
+	if ctx.Err() != nil {
+		t.Fatal("Context expired before RPCs failed.")
+	}
+}
+
+// This helper creates a management server for the test.
+func startManagementServer(t *testing.T) *e2e.ManagementServer {
+	t.Helper()
+	ms, err := e2e.StartManagementServer(e2e.ManagementServerOptions{})
+	if err != nil {
+		t.Fatalf("Failed to start management server: %v", err)
+	}
+	t.Cleanup(ms.Stop)
+	return ms
+}
+
+// This helper generates a custom bootstrap config for the test.
+func generateBootstrapContents(t *testing.T, serverURI string, ignoreResourceDeletion bool) []byte {
+	t.Helper()
+	bootstrapContents, err := bootstrap.Contents(bootstrap.Options{
+		NodeID:                             nodeID,
+		ServerURI:                          serverURI,
+		ServerListenerResourceNameTemplate: e2e.ServerListenerResourceNameTemplate,
+		IgnoreResourceDeletion:             ignoreResourceDeletion,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	return bootstrapContents
+}
+
+// This helper generates a custom bootstrap config for the test.
+func generateCustomResolver(t *testing.T, bootstrapContents []byte) resolver.Builder {
```

Comment:
Hmm ... you are not really *generating* a resolver, and this is not a *custom* resolver. So, maybe something like `buildXDSResolver()` would be better?

Also, should you be calling `t.Cleanup(r.Close())` here in this function?

Also, mark the function with `t.Helper()`.

And s/cResolver/r.

Reply:
You are right! This helper actually creates the builder. Renaming this to `xdsResolverBuilder` made more sense to me. Let me know your thoughts. 

> Also, should you be calling t.Cleanup(r.Close()) here in this function?

I think r.Close() is on the resolver and not the builder. I think the resolver is build()-ed and close()-ed inside of DIal. Let me know if i'm in the wrong here. 

>Also, mark the function with t.Helper().
And s/cResolver/r.

done

Reply:
Oh yes, you are right :)
I got confused with the case where we build manual resolvers and in that case, we usually close the resolver.